man/hub.1
.\" ** You probably do not want to edit this file directly **
.\" It was generated using the DocBook XSL Stylesheets (version 1.69.1).
.\" Instead of manually editing it, you probably should edit the DocBook XML
.\" source for it and then use the DocBook XSL Stylesheets to regenerate it.
.TH "HUB" "1" "12/13/2011" "\ " "\ "
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
hub \- manage Haskell hubs (for selecting toolchains, libraries and sandboxes)
.SH "SYNOPSIS"
.sp
.nf
    hub usage
    hub help    [<hub\-command>]
    hub version
    hub default [<g\-hub>|\-]
    hub ls
    hub set     [<hub>|\-]
    hub info    [<hub>]
    hub name
    hub path    [<hub>]
    hub xml     [<hub>]
    hub init    [<hub>]   <u\-hub'>
    hub cp      [<u\-hub>] <u\-hub'>
    hub mv      [<u\-hub>] <u\-hub'>
    hub rm       <u\-hub>
    hub swap    [<u\-hub>] <u\-hub'>
.fi
.SH "DESCRIPTION"
Each of the installed Haskell toolkit commands vector through the hub wrapper for selecting the apropropriate compiler tools and libraries (i.e., package database). The hub command is used to manage and configure the named \fIhubs\fR that select the versions of the tools and libraries to use. These hubs can be used as flexible, lightweight development sandboxes.
.sp
Each hub is named. If the name starts with a numeric digit then it is a global hub \(em being shared, imutable and preinstalled (with the system packagemanager). Otherwise it names a (private) user hub that can have packages installed into it.
.sp
Use \fIhub ls\fR to list the available hubs and \fIhub init\fR to create a new hub based on an existing hub \(em the new hub will inherit all of the global packages of the hub it is based on.
.sp
To mark a directory and its subdirectories for use with a hub use \fIhub set\fR. To invoke a toolkit command in the context of a particular hub set the HUB environment variable for the execution of the command. E.g.,
.sp
.sp
.nf
      HUB=2011.2.0.1 ghc\-pkg list
.fi
To copy, rename, remove and swap the contents of two hubs use \fIhub cp\fR, \fIhub mv\fR, \fIhub rm\fR and \fIhub swap\fR, respectively.
.sp
.SH "COMMANDS"
.TP
hub usage
Lists the syntax of all the hub commands.
.TP
hub help [<hub\-command>]
Lists the help for a command or all commands if none specified. See "hub usage" for a command\-syntax summary.
.TP
hub version
Lists the version information.
.TP
hub default [<g\-hub>|\-]
(Needs to be run as root.) If no arguments are given then this command lists the the default global hub for the system (i.e., the default global hub used to set up each user\(cqs
\fIhome\fR
hub). If a global hub <g\-hub> is specified then <g\-hub> will become the default global hub. If a
\fI\-\fR
is specified then any older default settings are discarded and the system default re\-established.
.TP
hub ls
Lists your user hubs and all of the global hubs.
.TP
hub set [<hub>|\-]
Sets the
\fIcurrent\fR
hub for a directory and its sub\-directories. The HUB environment variable can be set to a hub name to override this setting.
.TP
hub info [<hub>]
Lists the vital stats for the named or current hub. (See
\fIhub set\fR
on how to set the current hub.)
.TP
hub name
Lists the name of the current hub. (See
\fIhub set\fR
on how to set the current hub.)
.TP
hub path [<hub>]
Lists the path of the XML file defining the named or current hub. (See
\fIhub set\fR
on how to set the current hub.)
.TP
hub xml [<hub>]
Lists the contents of the XML file defining the named or current hub. (See
\fIhub set\fR
on how to set the current hub.)
.TP
hub init [<hub>] <u\-hub'>
Creates an empty user hub, <u\-hub'>. The new hub inherits the global hub of <hub> (or the current hub). (See
\fIhub set\fR
on how to set the current hub.)
.TP
hub cp [<u\-hub>] <u\-hub'>
Duplicates the user <u\-hub> (or the current hib) in <u\-hub'>. (See
\fIhub set\fR
on how to set the current hub.)
.TP
hub mv [<u\-hub>] <u\-hub'>
Renames user hub <u\-hub> (or the current hub) to <u\-hub'>. (See
\fIhub set\fR
on how to set the current hub.)
.TP
hub rm <u\-hub>
Deletes user hub <hub>.
.TP
hub swap [<u\-hub>] <u\-hub'>
Swaps the contents of user hub <u\-hub> (or the current hub) with user hub <u\-hub'>.
.SH "EXAMPLE 1"
List the packages in the (installed) 2011.2.0.1 Haskell Platform:
.sp
.sp
.nf
     HUB=2011.2.0.1 ghc\-pkg list
.fi
.SH "EXAMPLE 2"
Create a new user hub foo based on 2011.4.0.0 platform, attach it to the current directory, and install the latest edition of hexpat into it.
.sp
.sp
.nf
     hub init 2011.4.0.0 foo
     hub set foo
     cabal install hexpat
.fi
.SH "EXAMPLE 3"
Duplicates the 2010.4.0.0\-based foo hub of example 2 into hub bar and installs the SHA package into it. foo wil now contain the 2010.4.0.0 packages plus hexpat and any other packages it needs and bar will contain all this plus package SHA plus any additional packages it need.
.sp
.sp
.nf
     hub cp foo bar
     HUB=bar cabal install SHA
.fi
.SH "EXAMPLE 4"
Unpacks GLUT\-2.2.2.1 and runs one of its example programs using the (installed) Haskell Platform 2011.2.0.1.
.sp
.sp
.nf
     cabal unpack GLUT\-2.2.2.1
     HUB=2011.2.0.1 runhaskell GLUT\-2.2.2.1/examples/BOGLGP/Chapter01/Simple.hs
.fi
.SH "ENVIRONMENT VARIABLES"
.TP
HUB
This environment variable determines the hub that the hub wrapper will use while invoking a compiler tool. If it is not set then the .hub file (set by
hub set) in the current directory or one its parent directories is used to determine the
\fIcurrent\fR
hub. If no parent directory defines the current hub then the default global hub is used (see the
hub default
command). The
HUB
environment variable overrides all of these so it is best left unset unless you really want to force a single hub to be used for the current session regardless of current directory.
.SH "FILES"
.TP
[dot]hub
(A bug in docbook aparently forbids literals from starting with a period.) Lists the name of the hub for the current directory and it\(cqs subdirectories.
.TP
~/.hubrc/hub/<u\-hub>.xml
Contains the hub\-spefification file for user hub <u\-hub>.
.TP
~/.hubrc/lib/<u\-hub>/package.config
Contains the package database for user hub <u\-hub>.
.TP
/usr/hs/bin
Haskell tools ~\-\- this directory should be put on the user path if the tools aren\(cqt installed in /usr/bin.
.TP
/usr/hs/hub/<g\-hub>.xml
XML configuration file for global hub <g\-hub>.
.TP
/usr/hs/db
Package databases for system hubs.
.TP
/usr/hs/ghc/<version>
Installation directory for GHC <version>.
.TP
/usr/hs/hp/<version>
Installation directory for Haskell Platform <version>.
.SH "SEE ALSO"
.TP
hub(5)
Describes the format of the XML configuration files used to represent the user and global hubs.
.TP
[1]\&\fIhttp://justhub.org\fR
The home page of the JustHub distribution contains introductory material and links to wikis and articles.
.SH "BUGS"
The Hub issue tracker is at [2]\&\fIhttps://github.com/justhub/hub\-gen/issues\fR.
.sp
.SH "AUTHOR"
Chris Dornan <[3]\&\fIchris@chrisdornan.com\fR>
.sp
.SH "REFERENCES"
.TP 3
1.\ http://justhub.org
\%http://justhub.org
.TP 3
2.\ https://github.com/justhub/hub\-gen/issues
\%https://github.com/justhub/hub\-gen/issues
.TP 3
3.\ chris@chrisdornan.com
\%mailto:chris@chrisdornan.com


Hub/Hub.hs
module Hub.Hub
    ( HubName
    , Hub(..)
    , HubType(..)
    , homeHub
    , hubLib
    , sysVersion
    , sysDefaultHubPath
    , defaultHubPath
    , globalHubDir
    , hubBin
    , toolsBin
    , hubGccBin    
    , hubBinutilsBin
    , userHubDirs
    , allocate
    , hubUserLib
    , defaultGlobalHubName
    , hubExists
    , lsHubs
    , hubName
    , isGlobalHub
    , isGlobal
    , globalHubPath
    , checkHubName
    , userHubAvailable
    , userHubExists
    , userHubName
    , isUserHub
    , createHubDirs
    , userHubPath
    , userHubPaths
    , bin2toolchain
    , db2platform
    ) where

import qualified Control.Exception      as E
import           Data.Char
import           Data.List
import           Data.Maybe
import           System.Environment
import           System.Directory
import           System.FilePath
import           Text.Printf
import           Text.Regex
import           Hub.Oops
import           Hub.System


type HubName = String


data Hub = HUB {
    name__HUB :: HubName,
    path__HUB :: FilePath,
    hc_binHUB :: FilePath,
    tl_binHUB :: FilePath,
    glb_dbHUB :: FilePath,
    usr_dbHUB :: Maybe FilePath
    }                                                           deriving (Show)


data HubType
    = AnyHT
    | GlbHT
    | UsrHT
                                                                deriving (Show)

homeHub :: FilePath
homeHub = "home"

package_config :: FilePath
package_config = "package.config"

hubLib, sysVersion, sysDefaultHubPath, defaultHubPath, globalHubDir, hubBin,
                                toolsBin, hubGccBin, hubBinutilsBin :: FilePath
hc_bin_res, hp_db_res :: String
hubLib            = "/usr/hs/lib"
sysVersion        = "/usr/hs/lib/version.txt"
sysDefaultHubPath = "/usr/hs/lib/sys-default.hub"
defaultHubPath    = "/usr/hs/lib/the-default.hub"
globalHubDir      = "/usr/hs/hub"
hubBin            = "/usr/hs/bin"
toolsBin          = "/usr/hs/tools" 
hubGccBin         = "/usr/hs/gcc/bin"
hubBinutilsBin    = "/usr/hs/binutils/bin"
hc_bin_res        = "/usr/hs/ghc/([a-z0-9.-_]+)/bin"
hp_db_res         = "/usr/hs/db/(20[a-z0-9.-_]+)(.db)?/bin"

userHubDirs :: IO (FilePath,FilePath)
userHubDirs = 
     do hme <- home
        let hub =                       printf "%s/.hubrc/hub" hme
            lib =                       printf "%s/.hubrc/lib" hme
        return (hub,lib)

allocate :: IO FilePath
allocate =
     do hme <- home
        createDirectoryIfMissing True $ printf "%s/.hubrc/heap"             hme
        i <- inc                      $ printf "%s/.hubrc/heap/counter.txt" hme
        let pth =                       printf "%s/.hubrc/heap/%d"          hme i
        createDirectoryIfMissing True   pth
        return pth

user_lib :: FilePath -> HubName -> FilePath
user_lib hme hn = printf "%s/.hubrc/lib/%s" hme hn

db_re :: String -> Regex
db_re hme = mk_re $ printf "%s/.hubrc/lib/([^/]*)/%s/?" hme package_config

hn2fp :: HubName -> FilePath
hn2fp = (++ ".xml")

fp2hn :: FilePath -> Maybe HubName
fp2hn = match xml_fn_re

hubUserLib :: Hub -> IO (FilePath,FilePath)
hubUserLib hub = 
     do hme <- home
        case usr_dbHUB hub of
          Nothing -> oops SysO "no user DB speceified for this hub"
          Just db -> case match (db_re hme) db of
                       Just hn | hn==name__HUB hub
                            -> return $ (user_lib hme hn,db)
                       _    -> oops SysO "hub has non-standard user-database path"

xml_fn_re :: Regex
xml_fn_re = mk_re "(.*)\\.xml"



defaultGlobalHubName :: IO HubName
defaultGlobalHubName = sel
        [ usr_df     -- user-spec default
        , sys_df     -- system    default (may not be installed)
        , lhp_df     -- looks like the latest H.P. (guess)
        , lst_df     -- lexigographical maximum 
        ]
      where
        sel []        = oops SysO "no global hubs!"
        sel (p:ps)    =
                 do ei <- tryIO $ trim `fmap` p
                    case ei of
                      Left  _  -> sel ps
                      Right hn -> 
                         do ex <- fileExists $ globalHubPath hn
                            case ex of
                              True  -> return hn
                              False -> sel ps

        usr_df        = readAFile defaultHubPath

        sys_df        = readAFile sysDefaultHubPath

        lhp_df        = mx_fr $ gmatch hp_hub_re
        
        lst_df        = mx_fr $ const True

        mx_fr  p      = (filter p `fmap` lsHubs GlbHT) >>= mx 

        mx     []     = ioError $ userError "Hub.Hub: internal error"
        mx     (x:xs) = return  $ foldl max x xs
                
        hp_hub_re     = mk_re "20[0-9][0-9]\\.[0-9]\\.[0-9]\\.[0-9]"

hubExists :: HubName -> IO ()
hubExists hn =
     do hf <- case isGlobal hn of
                True  -> return $ globalHubPath hn
                False -> userHubPath hn
        ok <- fileExists hf
        case ok of
          True  -> return ()
          False -> oops SysO $ printf "%s: no such hub" hn

lsHubs :: HubType -> IO [HubName]
lsHubs GlbHT = ls_glb_hubs
lsHubs UsrHT = ls_usr_hubs
lsHubs AnyHT =
     do g_hns <- ls_glb_hubs
        u_hns <- ls_usr_hubs 
        return $ g_hns ++ u_hns

ls_glb_hubs :: IO [HubName]
ls_glb_hubs = (sort . chk) `fmap` getDirectoryContents globalHubDir
      where
        chk fps = [ hn | fp<-fps, Just hn<-[fp2hn fp], is_hub_name GlbHT hn ]

ls_usr_hubs :: IO [HubName]
ls_usr_hubs = 
     do dp <- fst `fmap` userHubDirs
        (sort . chk) `fmap` getDirectoryContents dp
      where
        chk fps = [ hn | fp<-fps, Just hn<-[fp2hn fp], is_hub_name UsrHT hn ]




hubName :: Hub -> HubName
hubName hub = name__HUB hub

isGlobalHub :: Hub -> Bool
isGlobalHub = isGlobal . name__HUB

isGlobal :: HubName -> Bool
isGlobal (c:_) = isDigit c
isGlobal _     = False

globalHubPath :: HubName -> FilePath
globalHubPath hn = globalHubDir </> hn2fp hn


-- tests for presence of Hub

checkHubName :: HubType -> HubName -> IO ()
checkHubName ht hn
    | is_hub_name ht hn = return ()
    | otherwise         = oops SysO $ printf "%s: invalid %shub name" hn ht_s
      where
        ht_s =  case ht of
                  AnyHT -> ""
                  GlbHT -> "global "
                  UsrHT -> "user "

is_hub_name :: HubType -> HubName -> Bool
is_hub_name ht hn =
        case hn of
          c:cs -> fst_hubname_c ht c && all (hubname_c ht) cs
          _    -> False

userHubAvailable :: HubName -> IO ()
userHubAvailable hn = 
     do userHubName hn
        iuh <- isUserHub hn
        case iuh of
          True  -> oops SysO $ printf "%s: hub already in use" hn
          False -> return ()

userHubExists :: HubName -> IO ()
userHubExists hn = 
     do userHubName hn
        hubExists   hn

userHubName :: HubName -> IO ()
userHubName hn =
        case isGlobal hn of
          False -> return ()
          True  -> oops SysO $ printf "%s: not a user hub" hn



isUserHub :: HubName -> IO Bool
isUserHub hn = userHubPath hn >>= fileExists


createHubDirs :: IO ()
createHubDirs =
     do (hub,lib) <- userHubDirs 
        createDirectoryIfMissing True hub
        createDirectoryIfMissing True lib

userHubPath :: HubName -> IO FilePath
userHubPath hn = (\(h_fp,_,_)->h_fp) `fmap` userHubPaths hn

userHubPaths :: HubName -> IO (FilePath,FilePath,FilePath)
userHubPaths hn = 
     do (hub,lib) <- userHubDirs
        let h_l = lib </> hn
        return (hub </> hn2fp hn, h_l, h_l </> package_config)


fst_hubname_c, hubname_c :: HubType -> Char -> Bool
fst_hubname_c AnyHT c = glb_first_hub_name_c c || usr_first_hub_name_c c
fst_hubname_c GlbHT c = glb_first_hub_name_c c
fst_hubname_c UsrHT c = usr_first_hub_name_c c
hubname_c     _     c = c `elem` "_-." || isAlpha c || isDigit c

glb_first_hub_name_c, usr_first_hub_name_c :: Char -> Bool
glb_first_hub_name_c c = isDigit c
usr_first_hub_name_c c = c `elem` "_." || isAlpha c



bin2toolchain, db2platform :: FilePath -> Maybe String
bin2toolchain = match $ mk_re hc_bin_res
db2platform   = match $ mk_re hp_db_res



mk_re :: String -> Regex
mk_re re_str = mkRegexWithOpts (printf "^%s$" re_str) False True

gmatch :: Regex -> String -> Bool
gmatch re st = isJust $ matchRegex re st

match :: Regex -> String -> Maybe String
match re st = case matchRegex re st of
                Just (se:_) -> Just se
                _           -> Nothing

home :: IO FilePath
home = catchIO (getEnv "HOME") $ \_ -> return "/"

tryIO :: IO a -> IO (Either IOError a)
tryIO = E.try

catchIO :: IO a -> (IOError->IO a) -> IO a
catchIO = E.catch
